docs/developers/runtime_abstraction_layer.md

```diff
@@ -0,0 +1,208 @@
+whole lifetime of the context in order to achieve the best performance. Based on
+the initialization-free interfaces (e.g.
+load_kernel_if_not_and_then_launch_kernel) provided by RAL, compiler side can
+focus on its core optimization logic and lets the RAL to manage the resource
+status.
+
+# 2 Design Overview
+
+## 2.1 From the Perspective of Compiler Side
+
+In the compiler side, RAL has a set of transformation passes to make sure the
+compiled binary is suitable for the RAL runtime.
+
+### 2.1.1 Context Injection
+
+As mentioned above，we prefer to simplify the design of the compiler side and
+let it foucus on its core optimization logic. Thus, RAL provides a context
+object to hide all stateful operations behind the context. The context is passed
+as a parameter to the entry function of the compilation module and all RAL APIs
+should always use the context as their first argument. RAL uses a context
+injection pass to ensure this property. The pass rewrites the entry function and
+all related functions to make sure their first argument is the context. To be
+concrete, we create a custom dialect disc_ral using MLIR infra to model the ral
+runtime behaviour. Inside the dialect, we define a custom type
+disc_ral.RalExecutionContextType to represent the context type in the compiler
+side. Under the hood, the disc_ral.RalExecutionContextType will be lowered to a
+pointer in llvm IR.
+
+### 2.1.2 Inputs/Outputs Binding
+
+For the entry function, RAL also rewrites its inputs and outputs to make it
+suitable to interact with different host environments. To be concrete, all the
+original inputs and outputs of the entry function are received from and sent to
+RAL through a sequence of RAL API calls correspondingly. The motivation behind
+this is to hide the implementation details of memref structure to make the ABI
+clean and stable.  Below is an example IR before and after the transformation.
+
+Original IR:
+
+```
+func @main(%arg0 : memref<?x?xf32>, %arg1 : memref<?x?xf32>) -> memref<?x?xf32> {
+  %ret = alloc(...)
+  use(%arg0, %arg1, %ret, ...)
+  return %ret : memref<?x?xf32>
+}
+```
+
+After conversion:
+
+```
+func @main(!disc_ral.context %ctx) {
+  %arg0 = disc_ral.recv_input(%ctx, 0) // receive the first input
+  %arg1 = disc_ral.recv_input(%ctx, 1) // receive the second input
+  %ret = alloc(...)
+  use(%arg0, %arg1, %ret, ...)
+  disc_ral.send_output(%ctx, 0, %ret) // send the first outpt
+}
+```
+
+As shown in the above example, the inputs (similiar to the outputs) are received
+from the ral context one by one. We could choose to use one api call to receive
+all inputs at the same time, which may also have lower runtime overhead in terms
+of the number of ral api calls. However, the current design lets the compiler
+have the opportunity the adjust the placement of each recv_input and send_output
+api to achieve partial execution of a compiled binary before all its inputs are
```

Comment:
Nit. Better to give a small example here. Partial execution is a nice feature, but maybe a little bit against normal intuition.